File: calc_e2e_time.py
#!python3

from datetime import datetime
import re
import requests

url_tpl = 'https://internal.pingcap.net/idc-jenkins/blue/rest/organizations/jenkins/pipelines/tidb-operator-pull-e2e-kind/runs/{job_num}/log/?start=0'

def parse_time(text):
    time_str = re.match(r'\[(.+)\]', text).group(1)
    time = datetime.strptime(time_str, '%Y-%m-%dT%H:%M:%S.%fZ')
    print(time)
    return time

def local_pv_time():
    job_num = 3632
    url = url_tpl.format(job_num=job_num)
    print(url)
    log = requests.get(url).text.splitlines()
    print('log downloaded')
    start_str = 'info: waiting for pods of daemonset kube-system/local-volume-provisioner are ready'
    end_str = 'info: all pods of daemonset kube-system/local-volume-provisioner are ready'
    for line in log:
        if start_str in line:
            start_time = parse_time(line)
            continue
        if end_str in line:
            end_time = parse_time(line)
            continue
    print('installing local-volume-provisioner takes {time}'.format(time=(end_time-start_time)))

if __name__ == '__main__':
    # TODO: parse start, end, static section, test section
    local_pv_time()

File: calc_e2e_time.js
const fs = require('fs');
const puppeteer = require('puppeteer');

const DATA_DIR = 'e2e_time_data'

async function extractPageContent(page) {
  const startTime = await page.evaluate(() => {
    let timeElems = document.querySelectorAll('.RunDetailsHeader-times div')
    let durationStr = timeElems[0].querySelector('span').innerText
    let duration = parseDurationToSeconds(durationStr)
    let startTimeStr = timeElems[1].querySelector('time').getAttribute('datetime')
    let startTimestamp = Date.parse(startTimeStr)
    return { durationStr, duration, startTimestamp, startTimeStr }
  })

  const failedTests = await page.evaluate(() => (
    [...document.querySelectorAll('.result-item.failure')].map(extractTest)
  ));

  const skippedTests = await page.evaluate(() => (
    [...document.querySelectorAll('.result-item.not_built')].map(extractTest)
  ));

  const passedTests = await page.evaluate(() => (
    [...document.querySelectorAll('.result-item.success')].map(extractTest)
  ));

  return {
    startTime,
    failedTests,
    skippedTests,
    passedTests
  }
}

async function extractPage(page, jobNum) {
  const url = `https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/tidb-operator-pull-e2e-kind/detail/tidb-operator-pull-e2e-kind/${jobNum}/tests/`
  await page.goto(url, { waitUntil: 'domcontentloaded' });
  try {
    await page.waitForSelector('.result-item', { timeout: 10000 })
  } catch {
    console.log('wait for .result-item element in page timeout, skip')
    return
  }
  // setup global helper functions in page js context, which is not in this nodejs context
  await page.evaluate(() => {
    window.extractTest = (elem) => {
      return {
        name: elem.querySelector('.result-item-label').innerText,
        durationStr: elem.querySelector('.result-item-extra-info').innerText,
        duration: parseDurationToSeconds(elem.querySelector('.result-item-extra-info').innerText)
      }
    }
    window.parseDurationToSeconds = (duration) => {
      if (duration === '<1s') {
        return 1
      }
      hms = duration.split(' ')
      sec = 0
      len = hms.length
      // hms.forEach(i => console.log(i))
      if (len > 0) {
        s = hms[len - 1]
        if (!s.endsWith('s')) {
          return -1
        }
        sec += parseInt(s)
      }
      if (len > 1) {
        m = hms[len - 2]
        if (!m.endsWith('m')) {
          return -2
        }
        sec += parseInt(m) * 60
      }
      if (len > 2) {
        h = hms[len - 3]
        if (!h.endsWith('h')) {
          return -3
        }
        sec += parseInt(h) * 3600
      }
      if (len > 3) {
        return -4
      }
      return sec
    }
  });

  console.log(url);
  page.on('console', msg => {
    console.log(msg.text())
  });
  let pageContent = await extractPageContent(page)
  await fs.writeFileSync(`${DATA_DIR}/${jobNum}.json`, JSON.stringify(pageContent, null, 2))
}

(async () => {
  const browser = await puppeteer.launch();
  const page = await browser.newPage();

  for (let jobNum = 1; jobNum < 3650; jobNum++) {
    console.log(`start extracting job ${jobNum}`)
    await extractPage(page, jobNum)
    console.log(`finish extracting job ${jobNum}`)
  }

  await browser.close();
})();
